### ISSUE TITLE
DuckStation's crashes possibly a MoltenVK issue?

### ISSUE THREAD

**Issue by tycooko** [KhronosGroup/MoltenVK#1196]
Hello,
My issue actually involves the DuckStation videogames' emulator (https://github.com/stenzek/duckstation). 
When I select "Renderer: Hardware (Vulkan)" inside DS's settings I can't choose the xBR texture filtering because, if I do, the whole software freezes and I have to force close it to use it again.
I mentioned this a couple of times to the creator, who, as in the following link, has repeatedly told me to research the issue within MoltenVK, since this isn't happening in Windows and he doesn't provide support for the Mac: https://github.com/stenzek/duckstation/issues/1217#issuecomment-744102985 .
I'm at kind of a dead end right now, because I don't know what to do to understand what's happening and how to solve it.
My OS is macOS Big Sur 11.1 and my computer is a 2020 27'' iMac.
Could you please help? Thanks!

**Comment by rcaridade145** [comment]
I am not familiar with the code base however those post processing filters seem to be defined /implemented on https://github.com/stenzek/duckstation/blob/master/src/frontend-common/vulkan_host_display.cpp and https://github.com/stenzek/duckstation/blob/e9e3891f9ec5d519cdb08c1b3c110f9c2de09a33/src/frontend-common/postprocessing_shadergen.cpp

**Comment by tycooko** [comment]
> I am not familiar with the code base however those post processing filters seem to be defined /implemented on https://github.com/stenzek/duckstation/blob/master/src/frontend-common/vulkan_host_display.cpp and https://github.com/stenzek/duckstation/blob/e9e3891f9ec5d519cdb08c1b3c110f9c2de09a33/src/frontend-common/postprocessing_shadergen.cpp

Hey, thanks for your reply! I'm afraid that pointing the team currently working on DuckStation in the right direction won't be enough, as, for what I've understood, none of them has a Mac and they can't do the necessary trial and error. 
Basically, what they want from me and anyone who wants this corrected is to give them the solution, already tried and confirmed.
I remember the creator of the software suggesting Vulkan might need an "update"?

Also, I don't know if this past issue with Dolphin might have anything to do: https://github.com/KhronosGroup/MoltenVK/issues/573 .

The DuckStation's team repeatedly asserted that DS is heavily based on Dolphin, like, they literally took most of the code from there.

**Comment by tycooko** [comment]
This is the (huge) report that gets automatically generated when DuckStation freezes and I force quit it:
[DuckStation report.txt](https://github.com/KhronosGroup/MoltenVK/files/5795105/DuckStation.report.txt)

**Comment by stenzek** [comment]
Based on that backtrace it looks like it's an issue with SPIRV-Cross. I dumped out the shaders it was compiling, and tried to compile with SPIRV-Cross externally, which gets stuck in the same stack (1).

I'm guessing this should be reported to SPIRV-Cross, as it's not directly a MoltenVK issue?

SPIR-V as well as the original GLSL attached below. While DuckStation's glslang version is a bit older, using the SPIRV-Cross version results in the same issue.

[shader_100.zip](https://github.com/KhronosGroup/MoltenVK/files/5797168/shader_100.zip)


(1)
```
#98 0x00005555555ad036 in exists_unaccessed_path_to_return(spirv_cross::CFG const&, unsigned int, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> > const&) ()
#99 0x00005555555ad036 in exists_unaccessed_path_to_return(spirv_cross::CFG const&, unsigned int, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> > const&) ()
#100 0x00005555555ad036 in exists_unaccessed_path_to_return(spirv_cross::CFG const&, unsigned int, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> > const&) ()
#101 0x00005555555ad036 in exists_unaccessed_path_to_return(spirv_cross::CFG const&, unsigned int, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> > const&) ()
#102 0x00005555555ae807 in spirv_cross::Compiler::analyze_parameter_preservation(spirv_cross::SPIRFunction&, spirv_cross::CFG const&, std::unordered_map<unsigned int, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> >, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<std::pair<unsigned int const, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> > > > > const&, std::unordered_map<unsigned int, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> >, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<std::pair<unsigned int const, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> > > > > const&) ()
#103 0x00005555555c002a in spirv_cross::Compiler::analyze_variable_scope(spirv_cross::SPIRFunction&, spirv_cross::Compiler::AnalyzeVariableScopeAccessHandler&) ()
#104 0x00005555555c1599 in spirv_cross::Compiler::build_function_control_flow_graphs_and_analyze() ()
#105 0x00005555556d7718 in spirv_cross::CompilerMSL::compile[abi:cxx11]() ()
#106 0x00005555555a2b4c in compile_iteration(CLIArguments const&, std::vector<unsigned int, std::allocator<unsigned int> >) ()
#107 0x00005555555a7318 in main_inner(int, char**) ()
#108 0x000055555559994a in main ()
```

**Comment by tycooko** [comment]
> Based on that backtrace it looks like it's an issue with SPIRV-Cross. I dumped out the shaders it was compiling, and tried to compile with SPIRV-Cross externally, which gets stuck in the same stack (1).
> 
> I'm guessing this should be reported to SPIRV-Cross, as it's not directly a MoltenVK issue?
> 
> SPIR-V as well as the original GLSL attached below. While DuckStation's glslang version is a bit older, using the SPIRV-Cross version results in the same issue.
> 
> [shader_100.zip](https://github.com/KhronosGroup/MoltenVK/files/5797168/shader_100.zip)
> 
> (1)
> 
> ```
> #98 0x00005555555ad036 in exists_unaccessed_path_to_return(spirv_cross::CFG const&, unsigned int, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> > const&) ()
> #99 0x00005555555ad036 in exists_unaccessed_path_to_return(spirv_cross::CFG const&, unsigned int, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> > const&) ()
> #100 0x00005555555ad036 in exists_unaccessed_path_to_return(spirv_cross::CFG const&, unsigned int, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> > const&) ()
> #101 0x00005555555ad036 in exists_unaccessed_path_to_return(spirv_cross::CFG const&, unsigned int, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> > const&) ()
> #102 0x00005555555ae807 in spirv_cross::Compiler::analyze_parameter_preservation(spirv_cross::SPIRFunction&, spirv_cross::CFG const&, std::unordered_map<unsigned int, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> >, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<std::pair<unsigned int const, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> > > > > const&, std::unordered_map<unsigned int, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> >, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<std::pair<unsigned int const, std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<unsigned int> > > > > const&) ()
> #103 0x00005555555c002a in spirv_cross::Compiler::analyze_variable_scope(spirv_cross::SPIRFunction&, spirv_cross::Compiler::AnalyzeVariableScopeAccessHandler&) ()
> #104 0x00005555555c1599 in spirv_cross::Compiler::build_function_control_flow_graphs_and_analyze() ()
> #105 0x00005555556d7718 in spirv_cross::CompilerMSL::compile[abi:cxx11]() ()
> #106 0x00005555555a2b4c in compile_iteration(CLIArguments const&, std::vector<unsigned int, std::allocator<unsigned int> >) ()
> #107 0x00005555555a7318 in main_inner(int, char**) ()
> #108 0x000055555559994a in main ()
> ```

Thank you very much for looking into it.

**Comment by HansKristian-Work** [comment]
@stenzek Fix is in flight, trivial bug.

**Comment by stenzek** [comment]
> @stenzek Fix is in flight, trivial bug.

Fantastic, thank you very much!

**Comment by billhollings** [comment]
Fixed in PR #1210. Please retest and close this issue if successful.
